modules/base/src/algorithm/mesh/meshclipping.cpp

```diff
@@ -0,0 +1,609 @@
+
+    // Find unique edges that surround the hole(s):
+
+    // Triangulation can get very bad very quickly, when clipping a mesh successively,
+    // i.e. we get very thin triangles. Depending on the value of EPSILON,
+    // too many edges are discarded, because they are found to be the same (result: missing faces!),
+    // or we cannot find closed edge loops anymore. When EPSILON is really small, we find loops with
+    // less than 3 edges, i.e. not polygons anymore.
+
+    // Possible solution: Find unique edges with low epsilon, find loops with high epsilon!
+
+    const auto uniqueintersectionsEdges = findUniqueEdges(intersectionsEdges, EPSILON);
+
+    if (uniqueintersectionsEdges.empty()) {
+        return outputMesh;
+    }
+
+    // Triangulate hole:
+
+    // With convex input we could assume only one hole, that is a convex, non-self-intersecting,
+    // closed polygon, but generally we have to find edge groups that form loops.
+
+    const auto loops = findLoops(uniqueintersectionsEdges, EPSILON);
+
+    // Simplify polygons, e.g. merging successive straight edges,
+    // to prevent triangulation becoming too bad, espescially when clipping a mesh multiple times.
+
+    const auto polygons = simplifyPolygons(loops, EPSILON);
+
+    // Calculate uv basis.
+    const auto u = vec3(polygons[0].get(0).v2 - polygons[0].get(0).v1);
+    const auto v = glm::cross(plane.getNormal(), u);
+
+    // Add new polygons as triangles to the mesh.
+    // Interpolating tex coords and color using barycentric coordinates.
+    for (size_t p = 0; p < polygons.size(); ++p) {
+        const size_t pSize = polygons[p].size();
+
+        std::vector<vec2> uv;
+        std::vector<vec3> tex;
+        std::vector<vec4> col;
+
+        for (size_t i = 0; i < pSize; ++i) {
+            // Calculate u-v plane coordinates of the vertex on the polygon
+            uv.push_back(
+                vec2(glm::dot(u, polygons[p].get(i).v1), glm::dot(v, polygons[p].get(i).v1)));
+
+            // Lookup texcoord and colors for the vertex of the polygon
+            for (size_t t = 0; t < intersectionTex.size(); ++t) {
+                if (intersectionTex.at(t).first == polygons[p].get(i).v1) {
+                    tex.push_back(intersectionTex.at(t).second);
+                    col.push_back(intersectionCol.at(t).second);
+                    break;
+                }
+            }
+        }
+
+        uv.pop_back();
+
+        std::vector<vec3> newVertices;
+        std::vector<vec3> newTexCoords;
+        std::vector<vec4> newColors;
+
+        // For polygons with less than 5 unique points, we could do better triangulation without the
+        // centroid, but the following works in any case.
+
+        // Calculate fake centroid
+        // (this will not give us the most optimal triangles,
+        // but it is sufficient to close the hole)
+        vec3 fakeCentroid(0.0f);
+        {
```

Comment:
Do you mean `(sum of points)/(number of points)`? It was my first thought too, but the result does not always lie on the plane.

I should also mention that there was a [previous (somewhat lengthy) code to find the real centroid](https://github.com/inviwo/inviwo/blob/6cadbb3e15476da324c0faa83847c8c8ed2ebcb4/modules/base/src/processors/meshclipping.cpp#L566). At one point during dev, when I saw bugs, I thought it might come from there, so I replaced it by something simpler, and I ended up keeping it, since it does the job.

Reply:
maybe Im missing something but why does (sum of points)/(number of points) not always lie on the plane? I use that in some other cases and I never had any problem. In my cases the polygon was always convex? Is that the reason? In the non convex case how is a centroid defined? Might be good with a comment explaining they the simple mean will not work...

Reply:
Sorry, you are right about the mean. It always lies on the plane, when one sums up only the unique points in the polygon, instead of the points of all edges, like I did. Will replace fakeCentroid by mean then.

Reply:
Ahh ic, did not see that you used the edge points... cool link, have never seen that one before!